### ISSUE TITLE
needless_range_loop invalid suggesstion with mut iterator and take

### ISSUE THREAD

**Issue by j-schwar** [rust-lang/rust-clippy#5945]
Hello! I found a little bug with the needless_range_loop lint. It's suggestion produces a compiler error when attempting to mutably iterate over part of a collection.

For example:

```rust
let mut vec = vec![1, 2, 3, 4];
for i in 0..vec.len() - 1 {
    vec[i] += 1;
}
println!("{:?}", vec);
```

Gives the following suggestion:

```
warning: the loop variable `i` is only used to index `vec`.
 --> src/main.rs:4:14
  |
4 |     for i in 0..vec.len() - 1 {
  |              ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(clippy::needless_range_loop)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_range_loop
help: consider using an iterator
  |
4 |     for <item> in vec.iter_mut().take(vec.len() - 1) {
  |    
```

The suggested change doesn't compile however:

```rust
for x in vec.iter_mut().take(vec.len() - 1) {
    *x += 1;
}
```

```
error[E0502]: cannot borrow `vec` as immutable because it is also borrowed as mutable
  --> src/main.rs:10:34
   |
10 |     for x in vec.iter_mut().take(vec.len() - 1) {
   |              ---            ---- ^^^ immutable borrow occurs here
   |              |              |
   |              |              mutable borrow later used by call
   |              mutable borrow occurs here
```

[Playground Link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=afd505dbcc246d2419ed0a2d4039e68e)

### Meta

- `cargo clippy -V`: clippy 0.0.212 (663d2f5cd 2020-08-22)
- `rustc -Vv`:
  ```
  rustc 1.47.0-nightly (663d2f5cd 2020-08-22)
  binary: rustc
  commit-hash: 663d2f5cd3163f17eddb74ee1e028d542255f21a
  commit-date: 2020-08-22
  host: x86_64-apple-darwin
  release: 1.47.0-nightly
  LLVM version: 10.0
  ```

**Comment by giraffate** [comment]
It seems to be same as https://github.com/rust-lang/rust-clippy/issues/1864.
